File: generated_code/GeminiPro/Java/065_SplayTree.java
/**
 * A self-contained Java class implementing a Splay Tree for an integer set.
 * Splay trees are self-balancing binary search trees with the additional property
 * that recently accessed elements are quick to access again.
 */
public class SplayTree {

    /**
     * Represents a node in the Splay Tree.
     */
    private static class Node {
        int key;
        Node left, right;

        /**
         * Constructor for a Node.
         * @param key The integer key for this node.
         */
        Node(int key) {
            this.key = key;
            this.left = null;
            this.right = null;
        }
    }

    private Node root;

    /**
     * Default constructor to create an empty Splay Tree.
     */
    public SplayTree() {
        this.root = null;
    }

    //
    // --- Private Helper Methods for Rotations ---
    //

    /**
     * Performs a right rotation on the subtree rooted at x.
     * @param x The root of the subtree to rotate.
     * @return The new root of the subtree after rotation.
     */
    private Node rightRotate(Node x) {
        Node y = x.left;
        x.left = y.right;
        y.right = x;
        return y;
    }

    /**
     * Performs a left rotation on the subtree rooted at x.
     * @param x The root of the subtree to rotate.
     * @return The new root of the subtree after rotation.
     */
    private Node leftRotate(Node x) {
        Node y = x.right;
        x.right = y.left;
        y.left = x;
        return y;
    }

    //
    // --- Splaying Operation ---
    //

    /**
     * The main splay operation. It brings the node with the given key
     * (or the last accessed node if the key is not found) to the root of the tree.
     * This is a recursive, bottom-up splay implementation.
     *
     * @param node The root of the current subtree.
     * @param key The key to search for and splay.
     * @return The new root of the tree after splaying.
     */
    private Node splay(Node node, int key) {
        // Base case: node is null or key is at the root of the subtree.
        if (node == null || node.key == key) {
            return node;
        }

        // Key is in the left subtree
        if (key < node.key) {
            // If left child is null, the key is not in the tree.
            if (node.left == null) {
                return node;
            }

            // Zig-Zig (Left-Left)
            if (key < node.left.key) {
                // Recursively bring the key to the root of the left-left grandchild's subtree.
                node.left.left = splay(node.left.left, key);
                // Perform the first rotation for the grandparent (current node).
                node = rightRotate(node);
            }
            // Zig-Zag (Left-Right)
            else if (key > node.left.key) {
                // Recursively bring the key to the root of the left-right grandchild's subtree.
                node.left.right = splay(node.left.right, key);
                // Perform the first rotation for the parent (node.left).
                if (node.left.right != null) {
                    node.left = leftRotate(node.left);
                }
            }

            // Perform the second rotation for the grandparent (or parent if it was a simple Zig).
            // A null check is needed in case the child became null after rotations.
            return (node.left == null) ? node : rightRotate(node);

        } else { // Key is in the right subtree
            // If right child is null, the key is not in the tree.
            if (node.right == null) {
                return node;
            }

            // Zig-Zag (Right-Left)
            if (key < node.right.key) {
                // Recursively bring the key to the root of the right-left grandchild's subtree.
                node.right.left = splay(node.right.left, key);
                // Perform the first rotation for the parent (node.right).
                if (node.right.left != null) {
                    node.right = rightRotate(node.right);
                }
            }
            // Zig-Zig (Right-Right)
            else if (key > node.right.key) {
                // Recursively bring the key to the root of the right-right grandchild's subtree.
                node.right.right = splay(node.right.right, key);
                // Perform the first rotation for the grandparent (current node).
                node = leftRotate(node);
            }

            // Perform the second rotation for the grandparent (or parent if it was a simple Zig).
            return (node.right == null) ? node : leftRotate(node);
        }
    }

    //
    // --- Public API Methods ---
    //

    /**
     * Searches for a key in the tree. If the key is found, it is splayed
     * to the root.
     *
     * @param key The key to search for.
     * @return The key if found, otherwise null.
     */
    public Integer search(int key) {
        root = splay(root, key);
        if (root != null && root.key == key) {
            return root.key;
        }
        return null;
    }

    /**
     * Inserts a key into the splay tree. If the key already exists,
     * the tree remains unchanged (as it's a set).
     *
     * @param key The key to insert.
     */
    public void insert(int key) {
        // If the tree is empty, create a new root.
        if (root == null) {
            root = new Node(key);
            return;
        }

        // Splay the tree for the key. This brings the closest node to the root.
        root = splay(root, key);

        // If the key is already at the root, it's a duplicate, so do nothing.
        if (root.key == key) {
            return;
        }

        // Create the new node.
        Node newNode = new Node(key);

        // The new node becomes the new root. The old root and its subtrees
        // are attached as children of the new root.
        if (key < root.key) {
            newNode.right = root;
            newNode.left = root.left;
            root.left = null;
        } else { // key > root.key
            newNode.left = root;
            newNode.right = root.right;
            root.right = null;
        }
        root = newNode;
    }

    /**
     * Deletes a key from the splay tree. If the key is not found,
     * the tree remains unchanged.
     *
     * @param key The key to delete.
     */
    public void delete(int key) {
        if (root == null) {
            return;
        }

        // Splay the key to the root.
        root = splay(root, key);

        // If the key is not at the root after splaying, it's not in the tree.
        if (root.key != key) {
            return;
        }

        // The node to delete is now the root.
        // We need to merge its left and right subtrees.
        Node leftSubtree = root.left;
        Node rightSubtree = root.right;

        if (leftSubtree == null) {
            // If there's no left subtree, the right subtree becomes the new tree.
            root = rightSubtree;
        } else {
            // Splay the left subtree to bring its maximum element to the root.
            // This maximum element will have no right child.
            leftSubtree = splay(leftSubtree, key); // or Integer.MAX_VALUE

            // Attach the original right subtree as the right child of the new root.
            leftSubtree.right = rightSubtree;
            root = leftSubtree;
        }
    }
}

File: generated_code/GeminiPro/Python/065_SplayTree.py
import sys

class SplayTree:
    """
    An implementation of a Splay Tree.

    This class provides a dictionary-like set for integers, supporting
    insert, delete, and search operations. The key feature is that any
    accessed (searched, inserted, or deleted) node is moved to the root
    of the tree through a series of rotations, a process called "splaying".
    This self-adjusting property provides good amortized performance for
    various access patterns.
    """

    class _Node:
        """A private helper class representing a node in the Splay Tree."""
        def __init__(self, key):
            self.key = key
            self.parent = None
            self.left = None
            self.right = None

    def __init__(self):
        """Initializes an empty Splay Tree."""
        self.root = None

    def _left_rotate(self, x):
        """Performs a left rotation on node x."""
        y = x.right
        x.right = y.left
        if y.left:
            y.left.parent = x

        y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        
        y.left = x
        x.parent = y

    def _right_rotate(self, x):
        """Performs a right rotation on node x."""
        y = x.left
        x.left = y.right
        if y.right:
            y.right.parent = x

        y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.right:
            x.parent.right = y
        else:
            x.parent.left = y

        y.right = x
        x.parent = y

    def _splay(self, node):
        """
        Performs the splaying operation on a node, moving it to the root.
        """
        while node.parent:
            parent = node.parent
            grandparent = parent.parent

            if not grandparent:  # Zig case
                if node == parent.left:
                    self._right_rotate(parent)
                else:
                    self._left_rotate(parent)
            elif node == parent.left and parent == grandparent.left:  # Zig-Zig (left-left)
                self._right_rotate(grandparent)
                self._right_rotate(parent)
            elif node == parent.right and parent == grandparent.right:  # Zig-Zig (right-right)
                self._left_rotate(grandparent)
                self._left_rotate(parent)
            elif node == parent.right and parent == grandparent.left:  # Zig-Zag (left-right)
                self._left_rotate(parent)
                self._right_rotate(grandparent)
            else:  # Zig-Zag (right-left)
                self._right_rotate(parent)
                self._left_rotate(grandparent)

    def insert(self, key):
        """
        Inserts an integer key into the tree.
        If the key already exists, the existing node is splayed to the root.
        If the key is new, it is inserted and then the new node is splayed.

        Args:
            key (int): The integer key to insert.
        """
        if not self.root:
            self.root = self._Node(key)
            return

        current = self.root
        parent = None
        while current:
            parent = current
            if key == current.key:
                # Key already exists, splay it and return
                self._splay(current)
                return
            elif key < current.key:
                current = current.left
            else:
                current = current.right

        # Insert the new node as a leaf
        new_node = self._Node(key)
        new_node.parent = parent
        if key < parent.key:
            parent.left = new_node
        else:
            parent.right = new_node

        # Splay the new node to the root
        self._splay(new_node)

    def search(self, key):
        """
        Searches for a key in the tree.
        Splays the found node to the root. If the key is not found, the last
        accessed node (the would-be parent) is splayed to the root.

        Args:
            key (int): The integer key to search for.

        Returns:
            bool: True if the key is found, False otherwise.
        """
        if not self.root:
            return False

        current = self.root
        last_visited = None
        while current:
            last_visited = current
            if key == current.key:
                self._splay(current)
                return True
            elif key < current.key:
                current = current.left
            else:
                current = current.right
        
        # Key not found, splay the last visited node
        if last_visited:
            self._splay(last_visited)
        return False

    def delete(self, key):
        """
        Deletes a key from the tree.
        The tree is first splayed on the key. If the key is found, it is
        removed, and its two subtrees are joined.

        Args:
            key (int): The integer key to delete.
        """
        if not self.root:
            return

        # Splay the tree on the key. This brings the node to delete (if it exists)
        # or its would-be parent to the root.
        found = self.search(key)

        # If the key is not at the root after splaying, it was not in the tree.
        if not found or self.root.key != key:
            return

        # The node to delete is now the root.
        left_subtree = self.root.left
        right_subtree = self.root.right

        if not left_subtree:
            # If there's no left subtree, the right subtree becomes the new tree.
            self.root = right_subtree
            if self.root:
                self.root.parent = None
        elif not right_subtree:
            # If there's no right subtree, the left subtree becomes the new tree.
            self.root = left_subtree
            if self.root:
                self.root.parent = None
        else:
            # Both subtrees exist. Join them.
            # 1. Find the maximum node in the left subtree.
            left_subtree.parent = None # Temporarily detach
            max_node = left_subtree
            while max_node.right:
                max_node = max_node.right

            # 2. Splay this max node to the root of the left subtree.
            #    This makes it the new root of the combined tree and ensures
            #    it has no right child, ready to adopt the right_subtree.
            self.root = left_subtree # Temporarily set root for splaying context
            self._splay(max_node)
            
            # 3. Attach the original right subtree as the right child of the new root.
            self.root.right = right_subtree
            right_subtree.parent = self.root